bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BufferedChannel.java

```diff
@@ -110,11 +150,15 @@ public long getFileChannelPosition() {
      * @throws IOException if the write or sync operation fails.
      */
     public void flush(boolean shouldForceWrite) throws IOException {
+        flush(shouldForceWrite, false);
+    }
+
+    public void flush(boolean shouldForceWrite, boolean forceMetadata) throws IOException {
         synchronized (this) {
```

Comment:
Looking at flushInternal(), that method should be synchronized. I would synchronize flushInternal, and rename it to flush(), then kill these two methods completely. Where they are being called, if the user wants to force, they should call forceWrite directly (should probably be renamed force too).

Reply:
yeah, it makes sense to add explicit 'synchronized' to flushInternal method, though it is currently called by the callers which are in synchronized block, so that it wont be misused by future callers.

Other than that, I want to make minimal changes to the code while adding the functionality I intend to add. I don't want to do code refactoring if it is not related to this change / caused by this change..

Reply:
+1 for minimal changes. since flushInternal is private, it won't be used outside of this class.

Reply:
I'm strongly -1 on adding another boolean parameter to this method. You're modifying now, so now is the time to do it cleanly.

Reply:
@ivankelly 

if you are breaking this method, you need to change all the places that use this method, which is effectively a "code refactor". what is the real point of doing this unrelated code refactor here, which sounds unrelated or not needed for the functionality here.

what is the real problem here about this method for you to give -1? a real performance issue, race condition? or just a coding style?

Reply:
boolean parameters are code smell. Code review isn't just about performance or race conditions.

Reply:
well, I don't think it is code smell. It is okay to me, because the flags define verify clear behavior: flush the buffer and if forceWrite enable, force the channel with 'forceMetadata' flag. The method follows the java standard: https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileChannel.html#force(boolean)

so going back to your comment here:

if your -1 is about the flag, `forceMetadata` flag is following the standard : https://docs.oracle.com/javase/7/docs/api/java/nio/channels/FileChannel.html#force(boolean)

if your -1 is about to ask Charan to break the methods into two methods, as Charan explained and I pointed out, this breakdown is a "code refactor" which will touch all the methods using `flush`. It is unrelated to the change, which can be done in a separate PR, so it doesn't sound to be a blocker to the change here.

so which part are you really giving -1 to?

Reply:
flushInternal just flushes writeBuffer to fileChannel (doesn't persists and moreover it is private method). forceWrite just forces filechannel (persists data written to filechannel so far, doesn't touches writeBuffer), flush calls flushInternal and provides option to whether do forceWrite or not.

So each method has different purpose. replacing flushinternal with flush is not correct thing to do from its expectations. And if user wants to do both flushing writeBuffer and forcewrites the filechannel, then calling just forceWrite is not right thing to do.

But again I don't see point in complicating things by changing the signatures of this commonly used methods and touch all the callers of them. It is extraneous.

Reply:
> well, I don't think it is code smell.

It's code smell. Google "boolean parameter". You'll get a selection of hits explaining why they are bad, including a blog post from Martin Fowler and another from Robert C Martin who wrote the book on clean code.

> because the flags define verify clear behavior

Do they? In 6 months are you going to read flush(false, true) and know exactly what the flags are doing?

> The method follows the java standard:

A 16 year old API. It can hardly be held up as a gold standard. Still though, it's not forceMetadata in itself I'm strongly opposed to. Its adding *another* boolean flag, to a method that already has one, that I'm opposed to.

> if your -1 is about to ask Charan to break the methods into two methods, as Charan explained and I pointed out, this breakdown is a "code refactor" which will touch all the methods using flush.

There's no refactor involved here. ```flush(boolean,boolean)``` isn't even called, so if it's removed, you will only have update ```flush(boolean)```.

Reply:
anyway, I would defer this to Charan. His code change works for me. I have also expressed my opinion here.

Reply:
> There's no refactor involved here. flush(boolean,boolean) isn't even called, so if it's removed, you will only have update flush(boolean).

It is not true. I'm planning to use in my next subtask. You can check how this method is used in my comprehensive PR - https://github.com/apache/bookkeeper/pull/1201 . We had quite big conversation regarding why 'forceMetadata' flag is required - (final conclusion of the conversation) - http://mail-archives.apache.org/mod_mbox/bookkeeper-dev/201705.mbox/%3CCAO2yDyahWu0ezJmQ66sdOvZBUofQu%3D8CabPDeH5JTRzkMULUCg%40mail.gmail.com%3E This conversation happened in May2017 for full thread - http://mail-archives.apache.org/mod_mbox/bookkeeper-dev/201705.mbox/browser

Reply:
refactor refers to existing code. The next subtask is not existing code, so calls in it to ```flush(boolean, boolean)```, can be replaced with calls to ```flush(); forceWrite(boolean);``` or ```flushAndForceWrite(boolean)```

Reply:
I was not referring to next subtask code for code refactoring. I was trying to explain -"flush(boolean,boolean) isn't even called", that it will be used by my future task and it cann't be removed.

It is not reasonable to expect people who touch the code intending to add new functionality to fix all the shortcomings of the existing code. It is risky and burdensome to touch critical path (change signatures) for mere arguable code cleanliness.

Reply:
>  fix all the shortcomings of the existing code.

You are adding a new method. I'm asking you not to add to issues that already exist.

> It is risky and burdensome to touch critical path (change signatures)

You don't have to change any signatures if you don't want to. Just call flush(false), and then forceWrite(true). flush() was just a suggestion as the false is redundant there. BufferedChannel should have a simple flush() operation that just flushing the buffer to the channel, which is exactly what flush(false) does. You don't have to add that though.